Run remove-virtual-functions on a per-function basis

smowton opened the pull request:
This introduces a mechanism to re-number a single GOTO program, since various parts of the CBMC codebase will assume the globally-unique-location-numbers invariant holds and it's safest to ensure it holds as much of the time as possible, then uses it to move remove-virtual-functions to run as each function is elaborated, rather than all in one go at the end.

smowton committed:
Introduce mechanism for renumbering an individual GOTO program

goto_programt::compute_location_numbers numbers it from zero, but of course these numbers
may clash with other programs; this adds the ability to renumber a program, at the cost of
wasting some "address" space.

tautschnig commented on src/goto-programs/goto_functions_template.h:
```diff
@@ -102,16 +102,23 @@ class goto_functions_templatet
   typedef goto_function_templatet<bodyT> goto_functiont;
   typedef std::map<irep_idt, goto_functiont> function_mapt;
   function_mapt function_map;
+  /// A location number such that unused_location_number - MAX_UINT are all
```
Is that `-` a minus (subtraction) or (more likely) indicating a range? Would you mind being clinical here and use a notation that unambiguously states whether that's a (half) open interval?

tautschnig commented on src/goto-programs/goto_functions_template.h:
```diff
@@ -102,16 +102,23 @@ class goto_functions_templatet
+  /// unused. There might still be unused numbers below this.
+  /// If numbering a new function or renumbering a function, starting from this
+  /// number is safe.
+  unsigned unused_location_number;
```
I don't think this should be `public`?

tautschnig commented on src/goto-programs/goto_model.h:
```diff
@@ -67,4 +67,28 @@ class goto_modelt
 };
 
+/// Interface for renumbering the locations in a goto_programt
+class goto_program_location_numberingt
```
What exactly is the value of this?

tautschnig commented on src/goto-programs/goto_model.h:
```diff
@@ -77,7 +77,7 @@ class goto_program_location_numberingt
   ///   goto programs, specifically incrementing its unused_location_number
   ///   member each time a program is re-numbered.
-  goto_program_location_numberingt(goto_modelt &goto_model):
+  explicit goto_program_location_numberingt(goto_modelt &goto_model):
```
Wrong commit. (Though in my earlier comment it doubted the overall value of this interface as well.)

tautschnig commented on src/goto-programs/lazy_goto_model.cpp:
```diff
@@ -23,14 +23,18 @@ lazy_goto_modelt::lazy_goto_modelt(
   post_process_functionst post_process_functions,
   message_handlert &message_handler)
   : goto_model(new goto_modelt()),
+    goto_model_numbering_interface(*goto_model),
```
Isn't this just a container for `goto_modelt`? I find this pretty weird.

smowton commented:
@tautschnig all comments addressed, except for the interface, whose purpose is to permit GOTO program renumbering (a non-const operation on `goto_functionst`) without providing a non-const reference to the entire function map, which should explicitly **not** be perturbed while post-processing a single function in isolation.

Alternatively solutions:
* Rewrite `location_number` as `union { struct { uint32_t function_number; uint32_t instruction_number; }; uint64_t location_number; }`, thus permitting GOTO program re-numbering without coordinating via `goto_functionst` again (good thing to do, more churny)
* Make `unused_location_number` mutable, so a const ref to `goto_functionst` suffices (don't like it; mutable is rather a blunt instrument compared to supplying a specific interface that exposes the required feature)
* Maintain a back-pointer from `goto_programt` to the function map it is in (error-prone; can a particular program (perhaps briefly) be in more than one map, for example?)

smowton committed:
JBMC: Run remove-virtual-functions as each function is converted

This means lazy loading drivers do not need to understand virtual
function calls, as they have already been converted into explicit
dispatch tables.

smowton commented:
@NathanJPhillips: please review (can't tag you due to your lack of write access to cbmc)

tautschnig commented:
> [...] purpose is to permit GOTO program renumbering (a non-const operation on `goto_functionst`) without providing a non-const reference to the entire function map, which should explicitly **not** be perturbed while post-processing a single function in isolation.

This sounds good to me, and certainly I wouldn't have complained at all had I read this comment in the source code. Would you mind adding it? ("Interface for renumbering the locations in a goto_programt" lacks the "why" that the above explanation does provide.)

tautschnig reviewed (approved):
Approving with the assumption that comments will be clarified (well, indeed they might just need to be shifted from the point-of-use to the definition).

smowton commented:
Added a comment as requested.

peterschrammel commented on src/goto-programs/goto_model.h:
```diff
@@ -67,4 +67,33 @@ class goto_modelt
+/// not be altered.
+class goto_program_location_numberingt
+{
+  goto_modelt &goto_model;
```
Nit pick: prefer protected members at bottom (as typical in the codebase).

peterschrammel commented on src/goto-programs/goto_program_template.h:
```diff
@@ -512,7 +514,10 @@ class goto_program_templatet
   {
     for(auto &i : instructions)
+    {
+      INVARIANT(nr != UINT_MAX, "Too many location numbers assigned");
```
use `std::numeric_limits::max`

peterschrammel commented on unit/testing-utils/load_java_class.cpp:
```diff
@@ -62,7 +62,9 @@ symbol_tablet load_java_class(
-    [] (goto_functionst::goto_functiont &function, symbol_tablet &symbol_table)
+    [] (goto_functionst::goto_functiont &function,
+        symbol_tablet &symbol_table,
+        goto_program_location_numberingt &location_numbering)
```
It feels quite odd nonetheless that we introduced `goto_modelt` as a container for goto_functions and symbol_table to simplify our interfaces and now here we are creating interfaces that require even more parameters. What about wrapping a goto_function and its goto_model into a `goto_function_modelt`? Then you can call renumber() on the goto_function_model without exposing the underlying goto_model.

smowton commented:
@peterschrammel you're quite right, I created `goto_model_functiont` which provides similar functionality but is much nicer. @tautschnig might want to look again since this has changed a bit.

tautschnig commented on src/goto-programs/goto_model.h:
```diff
@@ -67,4 +67,61 @@ class goto_modelt
+
+  /// Re-number our goto_function. After this method returns all instructions'
+  /// location numbers may have changed, but will be globally unique and in
+  /// program order within the program.
```
Nit pick: "program order" may be ambiguous (or possibly wrong) here; the numbers are strictly increasing as they are laid out in the underlying storage (source text or a list of statements), which need not coincide with the order of execution.

smowton replied:
This is the wording used elsewhere in `goto-programs/goto_program_template.h` so I'll keep this as-is

smowton commented:
Test-gen bump here: https://github.com/diffblue/test-gen/pull/1361

peterschrammel reviewed (approved):
This looks more elegant now.
